### PR TITLE
Task/user crud

### PR DESCRIPTION
**Pull request by marccolm**
#### What's this PR do?

Adds user crud functionality

#### Where should the reviewer start?
user component

#### How should this be manually tested?
In user section create, view, edit and delete users.

#### Any background context you want to provide?
Be kind, its my first time

This template was adapted ~stolen~ from [Quickleft/Sprint.ly](https://quickleft.com/blog/pull-request-templates-make-code-review-easier/)

### CODE REVIEW

File: src/app/admin/users/users.component.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                    .map((data: any) => data.json())
+                    .subscribe(userListJSON => {
+                      // Creates user objetc list from JSON.
+                      this.userList = [];
+                      for (let userJSON of userListJSON) {
+                        this.userList.push(new User(userJSON));
+                      }
+                    },
+                      err => {
+                        // Call of toast
```
Comment (JuanGro) [inline src/app/admin/users/users.component.ts]: Please, could you show a toaster with the error?

File: src/app/admin/users/users.component.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                    .map((data: any) => data.json())
+                    .subscribe(groupListJSON => {
+                      // Creates user objetc list from JSON.
+                      this.groupList = [];
+                      for (let groupJSON of groupListJSON) {
+                        this.groupList.push(new Group(groupJSON));
+                      }
+                    },
+                      err => {
+                        // Call of toast
```
Comment (JuanGro) [inline src/app/admin/users/users.component.ts]: Please, could you show a toaster with the error?

**Comment by JuanGro** [comment]
Please, fix this errors, you can see them typing 'ng lint'.

screen shot 2017-04-05 at 6 41 47 pm

I see that you have two errors in the tests, please fix them.

screen shot 2017-04-05 at 6 43 55 pm
screen shot 2017-04-05 at 6 43 49 pm

Check tests on client component, I think some tests are missing.

File: src/app/admin/users/users.component.ts (hunk 1 above)
Comment (marccolm) [inline src/app/admin/users/users.component.ts]: Done

File: src/app/admin/users/users.component.ts (hunk 2 above)
Comment (marccolm) [inline src/app/admin/users/users.component.ts]: Done

File: src/app/admin/users/user-form.component.html [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                    </div>
+                </div>
+            </div>
+        </div>
+
+
+        <!-- Fourth row -->
+        <div class="form-group row">
+            <div class="col-md-6">
+                <!-- Client field -->
```
Comment (EduardoVaca) [inline src/app/admin/users/user-form.component.html]: Not client field

File: src/app/admin/users/user-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    this.currentGroupList = [];
+    for (let group of object.groups_complete) {
+      let id: number = +group;
+      group = new Group();
+      group.id = id;
+      group.name = '';
+      this.currentGroupList.push(group);
+    }
+    object.groups_complete = this.currentGroupList;
+    console.log(object);
```
Comment (EduardoVaca) [inline src/app/admin/users/user-form.component.ts]: remove please

File: src/app/admin/users/user-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    if (this.user.id) {
+      // Update user
+      this.submitUpdatedUser(object, this.user.id);
+    } else {
+      // Create user
+      this.submitNewUser(object);
+    }
+    this.active = false;
+    setTimeout(() => this.active = true, 0);
+    this.success = true;
```
Comment (EduardoVaca) [inline src/app/admin/users/user-form.component.ts]: This variable is not needed. Remove it

File: src/app/admin/users/user-form.spec.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        it('should load correctly a user in user Input', () => {
+            component.user = testUser;
+            fixtureParent.detectChanges();
+            expect(component.user).toEqual(testUser);
+        });
+
+        /**
+        * Tests that the clients object list received from parent component is not empty.
+        **/
+        it('should load correctly clients list in clientsList Input', () => {
```
Comment (EduardoVaca) [inline src/app/admin/users/user-form.spec.ts]: Clients and clients list? This is users

File: src/app/admin/users/user-list.component.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+* Component which manage the table which shows the list of users:
+* - List all the users.
+* - Request open a modal depending of which option is selected.
+* - Send the user that is selected by the user.
+**/
+export class UserListComponent {
+  // Receives the user list from parent component.
+  @Input('userList') userList: User[];
+  // Receives the group list from parent component.
+  @Input('groupList') groupList: Group[];
```
Comment (EduardoVaca) [inline src/app/admin/users/user-list.component.ts]: Why do you need the group list in here?

File: src/app/admin/users/user-model.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    password?: string;
+    groups_complete?: Group[];
+    /**
+    * Constructor to create a User based on server response at POST.
+    * Params:
+    *   - jsonText: The user returned from API in json format string.
+    **/
+    constructor(json?) {
+      if (json) {
+      this.id = json && json.id;
```
Comment (EduardoVaca) [inline src/app/admin/users/user-model.ts]: Use indentation

File: src/app/admin/users/user-model.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      this.username = json && json.username;
+      this.first_name = json && json.first_name;
```
Comment (EduardoVaca) [inline src/app/admin/users/user-model.ts]: remove "json &&" you are already validating that the json is defined in the if(json)

File: src/app/admin/users/user-model.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+          this.groups_complete.push(new Group(group));
+        }
+      }
+      }
+    }
+}
+/**
+* Class that declares all the attributes used in the Group object.
+**/
+export class Group {
```
Comment (EduardoVaca) [inline src/app/admin/users/user-model.ts]: Separate Group class in another file. Doesn't matter it's small, it is supposed to be scalable... Maybe in the future extra stuff might be added to Group

File: src/app/admin/users/users.component.html [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+<div bsModal #newUserModal="bs-modal" class="modal fade" tabindex="-1" role="dialog" aria-labelledby="myNewUserModal" aria-hidden="true">
+    <div class="modal-dialog modal-lg" role="document">
+        <div class="modal-content">
+            <div class="modal-header">
+                <h4 class="modal-title">{{ titleNewModal }}</h4>
+            </div>
+            <!-- Common Form -->
+            <!-- Child component, it contains the user form -->
+            <!-- [user] sends the user empty to the form to avoid problems if the last user is currently saved -->
+            <!-- [clientsList] sends the clients list to the form -->
```
Comment (EduardoVaca) [inline src/app/admin/users/users.component.html]: Not clientList

File: src/app/admin/users/group-model.ts [hunk 12]
```diff
@@ -0,0 +1,16 @@
+/**
+* Class that declares all the attributes used in the Group object.
+**/
+export class Group {
+  id?: number;
+  name?: string;
+  /**
```
Comment (EduardoVaca) [inline src/app/admin/users/group-model.ts]: Leave blank line after name?:string;

File: src/app/admin/users/user-form.component.ts [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+  // Event for parent to update the currentUser.
+  @Output() userUpdated: EventEmitter<User> = new EventEmitter();
+  // Variable to check in test what action is executed between components.
+  public modalAction: string = '';
+  // Initialization of control form.
+  public userModalForm: FormGroup;
+  // Variable to return the old user if cancel the update form.
+  public oldUser: User;
+  // Variable to active the form.
+  public active: boolean = true;
```
Comment (EduardoVaca) [inline src/app/admin/users/user-form.component.ts]: this variable is not needed

File: src/app/admin/users/user-list.component.html [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                        <!-- Search field -->
+                        <!-- <div class="col-4 input-group align-self-end">
+                            <span class="input-group-addon"><i class="fa fa-search"></i>
+                            </span>
+                            <input type="text" class="form-control" placeholder="Search by name" />
+                        </div> -->
+                    </div>
+                </div>
+
+                <!-- Start table -->
```
Comment (EduardoVaca) [inline src/app/admin/users/user-list.component.html]: Make rows of table white (except the first one with titles)

File: src/app/admin/users/user-list.spec.ts [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    let fixtureChildTable: ComponentFixture<UserListComponent>;
+
+    // Save UsersComponent to test it's methods and variables.
+    let componentParent: UsersComponent;
+    // Save UserFormComponent to test it's methods and variables.
+    let componentForm: UserFormComponent;
+    // Save UserListComponent  to test it's methods and variables.
+    let component: UserListComponent;
+
+    // let httpServiceStub;
```
Comment (EduardoVaca) [inline src/app/admin/users/user-list.spec.ts]: If you are not using this delete it please

File: src/app/admin/users/user-model.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+* Class that declares all the attributes used in the User object.
+**/
+export class User {
+  id?: number;
+  username: string;
+  first_name: string;
+  last_name: string;
+  password?: string;
+  groups_complete?: Group[];
+  /**
```
Comment (EduardoVaca) [inline src/app/admin/users/user-model.ts]: Leave blank line after Group[];

**Comment by EduardoVaca** [comment]
Fix conflicts please

**Pull request merged**